### PR TITLE
crypto/tls: Fix a bug in the ECH extension

### PR DESCRIPTION
**Pull request by cjpatton**
To compute ClientHelloOuterAAD, the server unmarshals ClientHelloOuter,
removes the ECH extension, then marshal the result the same way the
client does in this package. As a result, the server may re-order and
remove unrecognized extensions in the processed. This changes the
computation of ClientHelloOuterAAD to operate on the serialized
ClientHelloOuter.

### CODE REVIEW

File: src/crypto/tls/ech.go [hunk 1]
```diff
@@ -882,6 +885,85 @@ func echIsValidVersion(ext uint16) bool {
 	return ext == extensionECH
 }
 
+// encodeClientHelloOuterAAD interprets data as ClientHelloOuter and maps it to
+// ClientHelloOuterAAD. Returns a bit indicated whether parsing ClientHelloOuter
```
Comment (chris-wood) [inline src/crypto/tls/ech.go:889]:
```suggestion
// ClientHelloOuterAAD. Returns a bool indicated whether parsing ClientHelloOuter
```

File: src/crypto/tls/ech.go [hunk 2]
```diff
@@ -882,6 +885,85 @@ func echIsValidVersion(ext uint16) bool {
...
+
+	if !s.ReadUint16(&legacyVersion) ||
+		!s.ReadBytes(&random, 32) ||
+		!s.ReadUint8LengthPrefixed(&legacySessionId) ||
+		!s.ReadUint16LengthPrefixed(&cipherSuites) ||
+		!s.ReadUint8LengthPrefixed(&legacyCompressionMethods) {
+		return nil, false
+	}
+
+	if !s.Empty() &&
```
Comment (chris-wood) [inline src/crypto/tls/ech.go:919]: Shouldn't extensions be non-empty? (Meaning, should fail if `s.Empty() == true` here?)

File: src/crypto/tls/ech.go [hunk 3]
```diff
@@ -882,6 +885,85 @@ func echIsValidVersion(ext uint16) bool {
...
+		(!s.ReadUint16LengthPrefixed(&extensions) || !s.Empty()) {
+		return nil, false
+	}
```
Comment (chris-wood) [inline src/crypto/tls/ech.go:922]: Should we assert that `s.Empty() == true` at this point? (We've reached the end of the CH.)

File: src/crypto/tls/handshake_client_tls13.go [hunk 4]
```diff
@@ -752,7 +753,7 @@ func (hs *clientHandshakeStateTLS13) abortIfRequired() error {
 	if c.ech.offered && !c.ech.accepted {
 		// If ECH was rejected, then abort the handshake.
 		c.sendAlert(alertECHRequired)
-		return errors.New("tls: ech: rejected")
+		return errors.New("tls: ech: rejected " + string(base64.StdEncoding.EncodeToString(c.ech.retryConfigs)))
```
Comment (chris-wood) [inline src/crypto/tls/handshake_client_tls13.go:756]: Why do we need to cast? `EncodeToString` returns a `string`, no?

**Review by chris-wood — APPROVED** [review]
LGTM with a few nits.

File: src/crypto/tls/ech.go (hunk 1 above)
Comment (cjpatton) [reply src/crypto/tls/ech.go:889]: Done, here and elsewhere in this file.

File: src/crypto/tls/ech.go (hunk 3 above)
Comment (cjpatton) [reply src/crypto/tls/ech.go:922]: Isn't that asserted above?
```
	if !s.Empty() &&
		(!s.ReadUint16LengthPrefixed(&extensions) || !s.Empty()) { // <--- ASSERTED AFTER "||"
		return nil, false
	}
```

File: src/crypto/tls/handshake_client_tls13.go (hunk 4 above)
Comment (cjpatton) [reply src/crypto/tls/handshake_client_tls13.go:756]: Oops, that's right. But I also didn't mean to commit this! Reverted.

File: src/crypto/tls/ech.go (hunk 3 above)
Comment (chris-wood) [reply src/crypto/tls/ech.go:922]: Oh, yeah. It's just not easy to read (for me). I'd recommend moving that to the following line to make it clear.

**Review by chris-wood — APPROVED** [review]

File: src/crypto/tls/ech.go (hunk 3 above)
Comment (cjpatton) [reply src/crypto/tls/ech.go:934]: Done.

**Pull request merged**